crates/shared/src/event_handling.rs

```diff
@@ -182,7 +197,23 @@ where
         if !have_deleted_old_events {
             self.store.replace_events(Vec::new(), range.clone()).await?;
         }
-        self.last_handled_block = Some(range.end().to_value());
+
+        track_block_range(&format!("range_{}", replacement_blocks.len()));
+        if !replacement_blocks.is_empty() {
+            // delete forked blocks
+            self.last_handled_blocks
+                .retain(|block| block.0 < replacement_blocks.first().unwrap().0);
+            // append new canonical blocks
+            self.last_handled_blocks
+                .extend(replacement_blocks.into_iter());
+            // cap number of blocks to MAX_REORG_BLOCK_COUNT
+            self.last_handled_blocks = self.last_handled_blocks[self
+                .last_handled_blocks
+                .len()
+                .saturating_sub(MAX_REORG_BLOCK_COUNT as usize)
+                ..self.last_handled_blocks.len()]
+                .to_vec();
```

Comment:
Nit, can we move the start end end computations out of the `[]` indexing? It would make it slightly more readable for me:
```suggestion
            let start_index = self
                .last_handled_blocks
                .len()
                .saturating_sub(MAX_REORG_BLOCK_COUNT as usize);
            self.last_handled_blocks = self.last_handled_blocks[start_index..]
                .to_vec();
```

Reply:
Fixed.